rskj-core/src/main/java/co/rsk/cli/migration/UnitrieMigrationTool.java

```diff
@@ -0,0 +1,396 @@
+            logger.error("Unitrie state root: {}", Hex.toHexString(lastStateRoot));
+            throw new IllegalStateException("Not matching state root");
+        } else {
+            logger.info("Migration complete");
+            logger.info("State root: {}", Hex.toHexString(lastStateRoot));
+        }
+
+        return unitrieRepository.getMutableTrie().getTrie();
+    }
+
+    private void buildPartialUnitrie(Trie orchidAccountsTrie, Repository repository) throws MissingContractStorageKeysException {
+        int accountsToLog = 500;
+        int accountsCounter = 0;
+        logger.trace("(x = {} accounts): ", accountsToLog);
+        Iterator<Trie.IterationElement> orchidAccountsTrieIterator = orchidAccountsTrie.getPreOrderIterator();
+        Collection<Keccak256> missingStorageKeys = new HashSet<>();
+        while (orchidAccountsTrieIterator.hasNext()) {
+            Trie.IterationElement orchidAccountsTrieElement = orchidAccountsTrieIterator.next();
+            TrieKeySlice currentElementExpandedPath = orchidAccountsTrieElement.getNodeKey();
+            if (currentElementExpandedPath.length() == Keccak256Helper.DEFAULT_SIZE) {
+                accountsCounter++;
+                byte[] hashedAddress = currentElementExpandedPath.encode();
+                OrchidAccountState oldAccountState = new OrchidAccountState(orchidAccountsTrieElement.getNode().getValue());
+                AccountState accountState = new AccountState(oldAccountState.getNonce(), oldAccountState.getBalance());
+                RskAddress accountAddress = addressHashes.get(ByteUtil.wrap(hashedAddress));
+                repository.createAccount(accountAddress);
+                repository.updateAccountState(accountAddress, accountState);
+                byte[] contractData = orchidContractDetailsDataStore.get(accountAddress.getBytes());
+                byte[] codeHash = oldAccountState.getCodeHash();
+                byte[] accountStateRoot = oldAccountState.getStateRoot();
+                if (contractData != null) {
+                    try {
+                        migrateContract(accountAddress, repository, contractData, codeHash, accountStateRoot);
+                    } catch (MissingContractStorageKeysException e) {
+                        missingStorageKeys.addAll(e.getMissingStorageKeys());
+                    }
+                }
+                if (accountsCounter % accountsToLog == 0) {
+                    logger.trace("x");
+                }
+            }
+        }
+        if (!missingStorageKeys.isEmpty()) {
+            throw new MissingContractStorageKeysException(missingStorageKeys);
+        }
+    }
+
+    private void migrateContract(RskAddress accountAddress, Repository currentRepository, byte[] contractData, byte[] accountCodeHash, byte[] stateRoot) throws MissingContractStorageKeysException {
+        ArrayList<RLPElement> rlpData = RLP.decode2(contractData);
```

Comment:
I would move all rlp parsing to another method, maybe there is a better way to split the code from this function

Reply:
Created a simple abstraction, but I wouldn't spend much time on something which will be removed soon.